### DISCUSSION TITLE
Best way for auto-generated credits page listing used software, plugins, etc.?

### DISCUSSION THREAD

**Discussion by Andre601** [mkdocs/mkdocs#3343] [Q&A] [answered]
I want to be as transparent as possible in terms of what tools and software I've used to make my blog, but keeping it updated manually seems like a really tedious task to do here, so I would like to automate this.

The question I do have now is, what would be the best aproach?
I want to have the page be appealing to look at. Maybe something like this:
https://paste.gg/p/Andre601/c5f9acc369144167b86754bfb6df4767

Preferably would I like to create the content in markdown format before it gets rendered, so that I don't have to deal with possible missing classes or class name changes.

Any ideas on what a good aproach could be?

**Comment by pawamoy** [comment]
Yes.

Here is my script to generate and render the data: https://github.com/mkdocstrings/griffe/blob/main/scripts/gen_credits.py. It picks up dependencies from pyproject.toml and pdm.lock (would work with poetry.lock as well I believe).

And here is how I include it in my docs: https://raw.githubusercontent.com/mkdocstrings/griffe/main/docs/credits.md. It is using [pymdownx.snippets](https://facelessuser.github.io/pymdown-extensions/extensions/snippets/) to include the code and [markdown-exec](https://github.com/pawamoy/markdown-exec) to execute it while building the docs.

Result: https://mkdocstrings.github.io/griffe/credits/

**Reply by Andre601** [reply]
Oh wow. That's quite cool. Will play around with it and let you know how it looks

**Reply by Andre601** [reply]
Sadly doesn't work. Gives me this exception: https://paste.helpch.at/pezoyanixe.php

Seems to be a missing dependency on TOML... Not sure how I would resolve that.

**Reply by oprypin** [reply]
Probably it needs `pip install toml`

**Reply by Andre601** [reply]
Yep, that resolved it. But now there is another issue 😅 

Now it seems like it wants a pyproject file for something that doesn't have one...
```shell
             Traceback (most recent call last):
               File "C:\Users\Andreas\AppData\Local\Programs\Python\Python311\Lib\site-packages\markdown_exec\formatters\python.py", line 59, in _run_python
                 exec(compiled, exec_globals)  # noqa: S102
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
               File "<code block: n1>", line 17, in <module>
                 pyproject = toml.load(project_dir / "pyproject.toml")
                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
               File "C:\Users\Andreas\AppData\Local\Programs\Python\Python311\Lib\site-packages\toml\decoder.py", line 133, in load
                 with io.open(_getpath(f), encoding='utf-8') as ffile:
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
             FileNotFoundError: [Errno 2] No such file or directory: 'pyproject.toml'
```

**Reply by Andre601** [reply, marked as answer]
I think I'll go with a very basic credits page for now... But still thank you for your help.

**Reply by oprypin** [reply]
This script is very specific, it assumes the project uses pyproject.toml and PDM build

**Reply by pawamoy** [reply]
Oh yes, it needs the `toml` library to read pyproject.toml and pdm.lock. And if you don't have those files, then it won't work. But you can very well adapt the script to either hardcode your dependencies, or fetch them from somewhere else (requirements.txt, etc.) :slightly_smiling_face:

**Reply by Andre601** [reply]
Well... It's not for a python project or alike... Simply for a blog where I want to credit the software used for it.

**Reply by pawamoy** [reply]
I see, well, all you need is to adapt the logic to fetch your credits data. The rest gives you an example of how to render it within your docs :slightly_smiling_face:
